### ISSUE TITLE
iOS 11: ActivationException triggered when registering a singleton

### ISSUE THREAD

**Issue by sbaechler** [simpleinjector/SimpleInjector#462]
In iOS 11 our app triggers an exception that it didn't in iOS 10 (using Xamarin):

The line

``` c#
container.RegisterSingleton<IService, RealService>();
```

triggers the exception:

> SimpleInjector.ActivationException: For the container to be able to create RealService it should have only one public constructor: it has none.

But the class `RealService` does have a public constructor.

Could there be a 32-bit only dependency?

I am using version 3.2.2.

**Comment by dotnetjunkie** [comment]
Few things:

- Can you show the code for `RealService`?
- Can you post the full stack trace?
- Can you upgrade to v3.3.2 and try again? What are the results?
- Can you upgrade to the latest v4.x release and try again? What are the results?

**Comment by sbaechler** [comment]
I updated to 4.0.11 but the error still persists. The following stack trace is with version 4.0.11. (Mono 5.2.0)

### RealService
```
    internal class RestManager : IRestManager
    {
        private readonly IPlatform platform;
        private readonly IRestService restService;

        public RestManager(IPlatform platform, IRestService restService)
        {
            this.platform = platform;
            this.restService = restService;
        }
        ... more methods here
}
```

### StackTrace
```
System.ArgumentException: For the container to be able to create RestManager it should have only one public constructor: it has none.
Parameter name: TImplementation
  at SimpleInjector.Container.ThrowArgumentExceptionWhenTypeIsNotConstructable (System.Type implementationType, System.String parameterName) [0x00010] in <eba7cb600f6e498d9e5aee18244f4416>:0
  at SimpleInjector.Container.Register[TService,TImplementation] (SimpleInjector.Lifestyle lifestyle, System.String serviceTypeParamName, System.String implementationTypeParamName) [0x0001b] in <eba7cb600f6e498d9e5aee18244f4416>:0
  at SimpleInjector.Container.RegisterSingleton[TService,TImplementation] () [0x00000] in <eba7cb600f6e498d9e5aee18244f4416>:0
  at PegmaSharedSyncLibrary.SyncLibrary.ConfigureDefault (SimpleInjector.Container container) [0x00001] in /Users/bao/projects/swiss/pegma-frontend/platforms/shared/PegmaSharedSyncLibrary/PegmaSharedSyncLibrary/SyncLibrary.cs:25
  at PegmaSharedSyncLibrary.SyncLibrary.Reset () [0x0001f] in /Users/bao/projects/swiss/pegma-frontend/platforms/shared/PegmaSharedSyncLibrary/PegmaSharedSyncLibrary/SyncLibrary.cs:48
  at PegmaSharedSyncLibrary.SyncLibrary..ctor () [0x00008] in /Users/bao/projects/swiss/pegma-frontend/platforms/shared/PegmaSharedSyncLibrary/PegmaSharedSyncLibrary/SyncLibrary.cs:15
  at PegmaSharedSyncLibrary.SyncLibrary.Instance () [0x00020] in /Users/bao/projects/swiss/pegma-frontend/platforms/shared/PegmaSharedSyncLibrary/PegmaSharedSyncLibrary/SyncLibrary.cs:66
  at PegmaIOS.AppController.ViewDidLoad () [0x0003a] in /Users/bao/projects/swiss/pegma-frontend/platforms/ios/PegmaIOS/AppController.cs:71
  at at (wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Users/builder/data/lanes/5024/152b654a/source/xamarin-macios/src/UIKit/UIApplication.cs:79
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x00038] in /Users/builder/data/lanes/5024/152b654a/source/xamarin-macios/src/UIKit/UIApplication.cs:63
  at PegmaIOS.Application.Main (System.String[] args) [0x00001] in /Users/bao/projects/swiss/pegma-frontend/platforms/ios/PegmaIOS/Main.cs:12
```

**Comment by dotnetjunkie** [comment]
It seems that something has changed in the Xamarin implementation of iOS 11,
X, compared to iOS 10. Where `typeof(RestManager).GetConstructors()` returns all `public` constructors of `internal` types (as it should), it seems that in iOS 11, `GetConstructors()` does not return any constructors if the parant type is `internal`.

Are you able to check this assumption?

**Comment by dotnetjunkie** [comment]
Correction: The .NET Standard 1.0 and 1.3 implementations of Simple Injector, use the following logic to find the public constructors of a type:

``` c#
type.GetTypeInfo().DeclaredConstructors.Where(ctor => !ctor.IsStatic && ctor.IsPublic); 
```

Assuming that you are using the .NET Standard 1.0 or 1.3 version of Simple Injector, it implies that the behavior of`MethodBase.IsPublic` changed between iOS 10 and iOS11. 

Would you mind checking what this expression returns for `RestManager` is both iOS 10 and iOS 11?

**Comment by bdurrer** [comment]
I was only able to check for iOS11:

`type.GetTypeInfo().DeclaredConstructors.Where(ctor => !ctor.IsStatic && ctor.IsPublic).Count();`, results in one match, which would be correct [I guess]. `MethodBase.IsPublic` is true

**Comment by dotnetjunkie** [comment]
Which version of Simple Injector are you using in your iOS11 project? netstandard1.0, netstandard1.3, net40 or net45?

**Comment by sbaechler** [comment]
We noticed that the linker removed all unused members of a class during compilation. Could it be that it removed our constructor since it was not directly used within our code? This would explain why the message said 'no constructor'.

**Comment by dotnetjunkie** [comment]
Well, if the linker removes constructors, than that could absolutely cause that problem.

Here are some ideas that might mitigate this:

- Make the type public. The linker might get less agressive.
- Create a bug report with the linker developers
- Refrain from using a DI Container and fall back to Pure DI.

This optimization mechanism reminds me of .NET Native, which does a similar thing. In .NET Native, reflection based libraries (like DI Containers) are practically unusable caused by the aggressive optimizations, cutting everything that the compiler/linker sees as unused.

With .NET Native I always advise the use of Pure DI instead of a Container, and this might hold in your case as well.

**Comment by TheBigRic** [comment]
Not sure if you use XForms or Xamarin Ios and if Xamarin Ios has the same options, but are you aware you can set the linker settings in the properties of the Ios project?

In my xamarin forms app, these look like:
![AltText](https://snag.gy/dOASPF.jpg)

**Comment by sbaechler** [comment]
We used a decorator to keep the members for other classes. Removing the linker would triple the size of the app. Unfortunately we cannot test this anymore since we have removed the DI library completely during the search for the issue. Maybe we'll put it back and test it again.

Anyway, I think I can close the issue since the bug was not within the library. Thank you very much for the quick response.
